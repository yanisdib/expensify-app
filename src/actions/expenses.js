import database from '../firebase/firebase';


// WITHOUT FIREBASE
// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// WITH FIREBASE INTEGRATION
// component calls action generator
// action generator returns function
// component dispatches function (?)
// function runs (has ability to dispatch other actions and whatever it wants)


// ADD_EXPENSE
// Takes an object as params
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

// Add an expense to Firebase database
// Takes an expense object as params
// Receives an empty object as default
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
// Receives an empty object as default params
// if no object is passed
export const removeExpense = ({ id } = {}) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    });

// EDIT_EXPENSE
export const editExpense = (id, updates) => (
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    });

// SET_EXPENSES
// Set the array value
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// Fetch Firebase database to retrieve existing expenses
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .orderByKey()
            .once('value').then((snapshot) => {
                const expenses = [];

                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
                console.log('Data has been successfully added to database!')
            });
    };
};


// Add an expense to Firebase database
// Takes a string and an object as params
export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
                console.log('Data has been successfully updated!')
            });
    };;
};

// Remove an expense from Firebase database
// Takes the id from a destructured object as params
// Receives an empty object as default
export const startRemoveExpense = ({ id = {} }) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
            console.log('Data successfully removed from database!');
        });
    };
};
