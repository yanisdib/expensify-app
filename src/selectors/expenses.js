/* eslint-disable array-callback-return */

import moment from "moment";

// Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; // if our expense's createdAt is smaller than startDate, it is filtered out.
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;     // if our expense's createdAt is greater or same as endDate, it is filtered out.
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};