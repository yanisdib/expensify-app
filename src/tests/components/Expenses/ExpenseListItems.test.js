import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItems from '../../../components/Pages/Expenses/ExpensesListItems';
import expenses from '../../fixtures/expenses';


test('should render ExpenseListItems correctly', () => {
    const wrapper = shallow(<ExpenseListItems {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});