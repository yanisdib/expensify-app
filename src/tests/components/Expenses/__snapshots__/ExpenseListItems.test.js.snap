// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItems correctly 1`] = `
<div>
  <[object Object]
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </[object Object]>
  <p>
    $195.00
     - 
    01/01/1970
  </p>
</div>
`;
